drake/systems/DrakeSystem.cpp

```diff
@@ -33,21 +36,59 @@ DrakeSystem::DrakeSystem(const std::string &_name, unsigned int num_continuous_s
           state_frame(make_shared<MultiCoordinateFrame>(name+"State",initializer_list<CoordinateFramePtr>({continuous_state_frame,discrete_state_frame}))),
           output_frame(make_shared<CoordinateFrame>(name+"Output",num_outputs,"y")) { }
 
-DrakeSystem::VectorXs DrakeSystem::getRandomState() {
-  return DrakeSystem::VectorXs::Random(state_frame->getDim());
+VectorXd DrakeSystem::getRandomState() {
+  return VectorXd::Random(state_frame->getDim());
 }
 
-DrakeSystem::VectorXs DrakeSystem::getInitialState() {
+VectorXd DrakeSystem::getInitialState() {
   return getRandomState();
 }
 
-void DrakeSystem::simulate(double t0, double tf, const DrakeSystem::VectorXs &x0, const SimulationOptions& options) const {
+DrakeSystemPtr DrakeSystem::tilqr(const Eigen::VectorXd& x0, const Eigen::VectorXd& u0, const Eigen::MatrixXd& Q, const Eigen::MatrixXd& R) {
+  if (!isTimeInvariant())
+  throw runtime_error("DrakeSystem::TILQR.  Time-invariant LQR only makes sense for time invariant systems.");
+
+  if (discrete_state_frame->getDim()>0)
+  throw runtime_error("DrakeSystem::TILQR.  Discrete-time LQR is not implemented yet (but would be easy).");
+  if (continuous_state_frame->getDim()<1)
+  throw runtime_error("DrakeSystem::TILQR.  This system has no continuous states.");
+
+  int num_x = continuous_state_frame->getDim(),
+      num_u = input_frame->getDim();
+
+  Eigen::VectorXd xu(num_x+num_u);
+  xu << x0, u0;
+  Drake::TaylorVecX xu_taylor = Drake::initTaylorVecX(xu);
+
+  auto xdot = autoDiffToGradientMatrix(dynamics(0*xu_taylor(1),xu_taylor.head(num_x),xu_taylor.tail(num_u)));
+  auto A = xdot.leftCols(continuous_state_frame->getDim());
+  auto B = xdot.rightCols(input_frame->getDim());
+
+  Eigen::MatrixXd K(num_u,num_x), S(num_x,num_x);
+  lqr(A, B, Q, R, K, S);
+
+  // todo: return the linear system with the affine transform.  But for now, just give the affine controller:
+  // u = u0 - K(x-x0)
+  #define nullmat MatrixXd::Zero(0,0)
+  #define nullvec VectorXd::Zero(0)
```

Comment:
well how about just
```
MatrixXd null_mat(0, 0);
VectorXd null_vec(0);
```
I know I said at some point that the ::Zero thing was fine, which it is, but this is more straightforward I guess. I thought I remembered there being an issue with just using those constructors but I was wrong.

But what would really improve readability is the addition of a couple of static methods for the creation of special types of AffineSystem, e.g
```
static AffineSystem AffineSystem::makeStaticGain(...)
static AffineSystem AffineSystem::makeContinuous(...)
static AffineSystem AffineSystem::makeDiscrete(...)
```